### ISSUE TITLE
Support for Request timeout in spring rest template

### ISSUE THREAD

**Issue by ChristianFinckler** [spring-projects/spring-boot#11379]
I use spring rest template for performing a head request with a proxy server.
The target host does not exist and the proxy just does not finish the request. The problem is, that neither the connect timeout nor the read timeout stop the processing.

My code:
```
final ResponseEntity<String> sourceRes = restTemplate.exchange(head(new URI(post.getSourceUrl())).build(), String.class);
        return sourceRes.getStatusCode().is2xxSuccessful();
```
Here is my code for setting up the resttemplate:
 ```
final HttpClientBuilder clientBuilder = HttpClientBuilder.create();
    clientBuilder.useSystemProperties();
    clientBuilder.setDefaultRequestConfig(custom()
      .setCookieSpec(STANDARD)
      .setConnectTimeout(Duration.ofSeconds(1).toMillis())
      .setSocketTimeout(Duration.ofSeconds(1).toMillis())
      .build());
    final HttpClient httpClient = clientBuilder.build();
    final HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory(httpClient);
    requestFactory.setConnectTimeout(Duration.ofSeconds(1).toMillis());
    requestFactory.setReadTimeout(Duration.ofSeconds(1).toMillis());

    final RestTemplate restTemplate = new RestTemplate(requestFactory);
```

Here is the request performed with curl:
```
 curl -X HEAD --proxy "proxyuser:proxypassword@proxyhost:1234t"  "https://iAmNotExisting.com/dummy.jpg" --verbose
* About to connect() to proxy proxyhost port (#0)
*   Trying 15.112.12.32...
* Connected to proxyhost (15.112.12.32) port 1234 (#0)
* Establish HTTP proxy tunnel to iAmNotExisting.com:443
* Proxy auth using Basic with user 'proxyuser'
> CONNECT iAmNotExisting.com:443 HTTP/1.1
> Host: iAmNotExisting.com:443
> Proxy-Authorization: Basic blablub=
> User-Agent: curl/7.29.0
> Proxy-Connection: Keep-Alive
> 
* Operation timed out after 300314 milliseconds with 0 out of 0 bytes received
* Closing connection 0
curl: (28) Operation timed out after 300314 milliseconds with 0 out of 0 bytes received
```

As you can see, curl sends the request headers, but does not receive any response.

I think, this is either a bug, because the socket timeout does not work or that a request timeout would be needed if the socket timeout is not intended for this use case

The behavior of the proxy could be caused, because it is a keep-alive session.

**Comment by bclozel** [comment]
Can you confirm that this does not happen when just using the raw `HttpClient`? If it still does, then you'd need to raise that issue to that project.

**Comment by ChristianFinckler** [comment]
This indeed happens also with the httpClient. I created an issue at their jira:
https://issues.apache.org/jira/browse/HTTPCLIENT-1892. 
At least the httpClient has an abort()-method. So I could implement a request timeout with my own extra thread and call abort after the timeout has reached. Lets see, what the apache guys say.

**Comment by bclozel** [comment]
Thanks for reporting that, @ChristianFinckler.
I've looked at it and it seems that both Spring Boot and Spring Framework do not set that socket timeout property. So you're not the only one who got it wrong...

Could you test your setup with a custom `HttpClient` like previously, but this time:

```java
SocketConfig socketConfig = SocketConfig.custom().setSoTimeout(Duration.ofSeconds(1).toMillis()).build();
final HttpClient client = HttpClients.custom().setDefaultSocketConfig(socketConfig).build();
```

If this works for you, then we might need to raise that as an enhancement request in both Framework and Boot to take that into account.

**Comment by ChristianFinckler** [comment]
Hello, After setting the defaulSocketConfig in my example above, it works. The timeout works the the indefinetly conecction setup gets aborted. 

As a feature request, it could be good, if requestFactory.setReadTimeout(Duration.ofSeconds(1).toMillis()); also sets the default socket timeout.

**Comment by bclozel** [comment]
The default socket config is set on the `HttpClient` itself; in our comments, we're using the default socket config but we can also configure that through the connection manager, and much more. I believe this goes hand in hand with the SSL configuration - which is available only on the `HttpClient` builder itself.

`RestTemplateBuilder` is definitely not at the right abstraction level - and even Spring Framework's `HttpComponentsClientHttpRequestFactory` only takes `HttpClient` as an immutable constructor argument (at that point it's too late).

I've created [SPR-16697](https://jira.spring.io/browse/SPR-16697) to at least provide a hint for developers using the {{setConnectTimeout}} method on the request factory. Hopefully this will save them some time if they're facing the same issue.

Thanks for the report!

**Comment by gandrade** [comment]
[SPR-16697](https://jira.spring.io/browse/SPR-16697) has mentioned `5.0.5` as an affected version. I'm using Spring Boot `1.5.4.RELEASE`, which uses Spring Framework `4.3.9` version, and noticed the same behavior. Can you guys confirm it? BTW, the code proposed works fine also.

**Comment by gandrade** [comment]
Just to point out, that running some tests, I've faced different timeout behaviors when using `OAuth2RestTemplate`. Does it make sense?

**Comment by gandrade** [comment]
Running some tests, I've noticed timeout configuration has being ignored when using `OAuth2RestTemplate`. Does it make sense?

**Comment by bclozel** [comment]
SPR-16697 was fixed in Spring Framework 5.0.5 and was not backported to 4.3.x.
Even if it was, this is just a documentation issue as we can’t apply that timeout at the request level but this needs to be done on the ˋHttpClient` instance itself.
So this is consistent with what you’re seeing right now.
